packages/axelar-wasm-std-derive/src/lib.rs

```diff
@@ -282,3 +282,71 @@ fn match_unit_variant(event_enum: &Ident, variant_name: &Ident) -> TokenStream2
         #event_enum::#variant_name => cosmwasm_std::Event::new(#event_name)
     }
 }
+
+#[proc_macro_attribute]
+pub fn migrate_from_version(input: TokenStream, item: TokenStream) -> TokenStream {
+    let base_version_req = syn::parse_macro_input!(input as syn::LitStr).value();
+    let annotated_fn = syn::parse_macro_input!(item as syn::ItemFn);
+
+    let fn_name = &annotated_fn.sig.ident;
+    let fn_inputs = &annotated_fn.sig.inputs;
+    let fn_output = &annotated_fn.sig.output;
+    let fn_block = &annotated_fn.block;
+
+    if fn_name != "migrate" {
+        return syn::Error::new(
+            fn_name.span(),
+            "#[migrate_from_version] can only be applied to a 'migrate' function",
+        )
+        .to_compile_error()
+        .into();
+    }
+
+    let deps = match deps_ident(&annotated_fn.sig) {
+        Ok(deps) => deps,
+        Err(e) => return e.to_compile_error().into(),
+    };
+
+    let gen = quote! {
+        pub fn #fn_name(#fn_inputs) #fn_output {
+            let contract_name = env!("CARGO_PKG_NAME");
+            let contract_version = env!("CARGO_PKG_VERSION");
+
+            let old_version = semver::Version::parse(&cw2::get_contract_version(#deps.storage)?.version)?;
+            let version_requirement = semver::VersionReq::parse(#base_version_req)?;
+            assert!(version_requirement.matches(&old_version), "{} contract migration from version {} to {} is not supported", contract_name, old_version, contract_version);
+
+            let result = (|| {
+                #fn_block
+            })();
+
+            cw2::set_contract_version(#deps.storage, contract_name, contract_version)?;
+
+            result
```

Comment:
```suggestion
            cw2::set_contract_version(#deps.storage, contract_name, contract_version)?;

            #fn_block
```
I see no point of executing the fn body itself as a closure call.

Reply:
I used the closure because that way I can execute the function body before setting the new version. We used to set the version at the end because sometimes we would have some others checks or tests inside the migration module that would depend on the previous version. Right now I'm not sure we will have something like that, but I still think it would be a good practice to leave setting the new version at the end of the execution, what do you think?